### ISSUE TITLE
hinted handoff: limiting memory consumption of hints waiting to be written to disk is broken

### ISSUE THREAD

**Issue by piodul** [scylladb/scylladb#17636]
HEAD: 0a7854ea4de04f20b71326ba5940b5fac6f7241a

In https://github.com/scylladb/scylladb/commit/6e79d647e6fd080af7441cc5c493e17ceb582aff, the behavior of `db::hints::manager::too_many_in_flight_hints_for` function was accidentally modified:

```diff
     // There is no need to check the DC here because if there is an in-flight hint for this
     // endpoint, then this means that its DC has already been checked and found to be ok.
     return _stats.size_of_hints_in_progress > MAX_SIZE_OF_HINTS_IN_PROGRESS
-            && !utils::fb_utilities::is_me(ep)
+            && _proxy.local_db().get_token_metadata().get_topology().is_me(ep)
             && hints_in_progress_for(ep) > 0
             && local_gossiper().get_endpoint_downtime(ep) <= _max_hint_window_us;
 }
```

Now, this function always returns `false` if the function is called on an IP (`ep` argument) different than the local node's IP. Because we never store hints for the local node, this basically means that there are never "too many hints" - i.e. we never return OverloadedException to the driver, but we can overload node's memory if the node doesn't keep up with writing hints to disk.

The fix is very simple - add the missing `!` before the modified expression.

Found by @gusev-p in https://github.com/scylladb/scylladb/pull/15567/commits/676b5a28ad935a506d9fc93b695298aa856abfa6#r1452549984

**Comment by nyh** [comment]
> @dawmd - do we need to backport the fix to any relevant branch?

https://github.com/scylladb/scylladb/commit/6e79d647e6fd080af7441cc5c493e17ceb582aff did not reach 5.4 so no backports are needed.

HOWEVER, I don't know how we can ensure that https://github.com/scylladb/scylladb/commit/6e79d647e6fd080af7441cc5c493e17ceb582aff won't be backported sometime in the future (because we suddenly realize we need it for something) - and then we'll need to backport this patch as well. I think the likelyhood of this is very low, but it seems a hole in our backport process that might bite us in the future, somewhere (probably not in this case).

**Comment by mykaul** [comment]
> > @dawmd - do we need to backport the fix to any relevant branch?
> 
> [6e79d64](https://github.com/scylladb/scylladb/commit/6e79d647e6fd080af7441cc5c493e17ceb582aff) did not reach 5.4 so no backports are needed.
> 
> HOWEVER, I don't know how we can ensure that [6e79d64](https://github.com/scylladb/scylladb/commit/6e79d647e6fd080af7441cc5c493e17ceb582aff) won't be backported sometime in the future (because we suddenly realize we need it for something) - and then we'll need to backport this patch as well. I think the likelyhood of this is very low, but it seems a hole in our backport process that might bite us in the future, somewhere (probably not in this case).

Indeed. Interesting hole: a test for this scenario would help, but it means that we'd need to also backport this test to an earlier version first...

**Comment by nyh** [comment]
> Indeed. Interesting hole: a test for this scenario would help, but it means that we'd need to also backport this test to an earlier version first...

This is one of the reasons why I suggested in the past that we should run the **latest** tests on old versions to detect missing backports (and white-list tests which we know fail on older versions and we're fine with them, e.g., because they are a new feature). In other words, if someone writes a new test for master, and it passes on master, but then we learn that it fails on Scylla 5.4, we need to ask ourselves (and make an explicit decision) if this is because we missed a backport, or it's a failure we want to allow (e.g., the new test checks a new feature that wasn't in 5.4).

Allowing such a process was one of the design goals of test/cql-pytest - you can run a new test on Scylla master, on an ancient Scylla or even on Cassandra, and it is also not very difficult to fix dtest/ccm to allow this as well. But the Jenkins "process" around this idea doesn't exist - currently we keep a separate branch for Scylla 5.4 tests (for example) and newly written tests never run against Scylla 5.4 - and we need to remember to backport tests as well, which is just as easy to forget doing as forgetting to backport the code.

**Comment by mykaul** [comment]
> > Indeed. Interesting hole: a test for this scenario would help, but it means that we'd need to also backport this test to an earlier version first...
> 
> This is one of the reasons why I suggested in the past that we should run the **latest** tests on old versions to detect missing backports (and white-list tests which we know fail on older versions and we're fine with them, e.g., because they are a new feature). In other words, if someone writes a new test for master, and it passes on master, but then we learn that it fails on Scylla 5.4, we need to ask ourselves (and make an explicit decision) if this is because we missed a backport, or it's a failure we want to allow (e.g., the new test checks a new feature that wasn't in 5.4).
> 
> Allowing such a process was one of the design goals of test/cql-pytest - you can run a new test on Scylla master, on an ancient Scylla or even on Cassandra, and it is also not very difficult to fix dtest/ccm to allow this as well. But the Jenkins "process" around this idea doesn't exist - currently we keep a separate branch for Scylla 5.4 tests (for example) and newly written tests never run against Scylla 5.4 - and we need to remember to backport tests as well, which is just as easy to forget doing as forgetting to backport the code.

How do we know it's not a new test for a new functionality, which obviously doesn't exist in older versions? We need to carefully mark the test as such, I reckon?

**Comment by nyh** [comment]
> How do we know it's not a new test for a new functionality, which obviously doesn't exist in older versions? We need to carefully mark the test as such, I reckon?

Yes, as I said, such a process will mean that every time a new test passes on master but fails on on an older version, there's a decision to be made if we're ok with that failure, and we need to mark this decision down somewhere (a tag in the test itself, or more likely an entry in a separate file, to support non-open-source branches). Basically it's the same decision we make for backports - is the test failing because it's a new feature we don't want to backport? Maybe it's a real bug but we decided it's too difficult to backport? But the failed test will "force" us to make a decision, it's not something we can forget to do.

By the way, even though we don't "officially" use such a process (running new tests on old branches) in our CI, I used it in the past in sporadic cases - e.g., to check whether we forgot to backport any Alternator patch to a certain version branch (by running the current Alternator test suite against that branch, and seeing what fails) before suggesting to a customer to use Alternator on that branch.
By the way, this is also one of the reasons that I insist when I review tests that a test should have a comment explaining what it reproduces, and when possible also refer to the issue it reproduces. This makes it easier, when we see this test fail, to know what we forgot to backport.

**Comment by mykaul** [comment]
BTW, it has a huge benefit that we do not need to maintain separate branches for our tests, inc. backports efforts, while we continue to improve coverage even for older releases. I think it's a good idea, though certainly has the downside cost of maintaining test code that would run across versions (and sometimes, APIs and behaviors change dramatically)

**Comment by nyh** [comment]
>  though certainly has the downside cost of maintaining test code that would run across versions (and sometimes, APIs and behaviors change dramatically)

This is a true point - it is possible that a new version deliberately changed an API or behavior, and the if we only keep the new test for the new API or behavior, we lose the testing for the old version. So we may want to keep around old variants of the tests, which isn't pretty.

By the way, this is the same problem we have right now when we changed the replication method (tablets vs vnodes) - we want to make sure the master tests check not only tablets but also vnodes (although in this case it's not because we want to test older branches - but because even the new branch has an option to use the older vnodes).
